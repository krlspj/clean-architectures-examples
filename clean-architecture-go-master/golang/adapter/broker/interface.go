package broker

type ProducerInterface interface {
	Publish(msg interface{}, key []byte, topic string) error
}
