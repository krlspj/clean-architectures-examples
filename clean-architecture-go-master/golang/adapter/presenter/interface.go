package presenter

type Presenter interface {
	Show() ([]byte, error)
	Bind(interface{}) error
}
