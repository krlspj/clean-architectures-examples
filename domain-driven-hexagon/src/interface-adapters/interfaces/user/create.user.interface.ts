/*  Creating interfaces like this may be useful if you need to share types with
    a front end web/mobile application, microservices, or other TypeScript APIs.
    You can share interfaces as a git submodule, a npm package, a library or in a monorepo, etc.
*/
export interface CreateUser {
  readonly email: string;
  readonly country: string;
  readonly postalCode: string;
  readonly street: string;
}
