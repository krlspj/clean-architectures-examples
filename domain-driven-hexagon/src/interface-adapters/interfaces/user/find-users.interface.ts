export interface FindUsers {
  readonly country: string;
  readonly postalCode: string;
  readonly street: string;
}
