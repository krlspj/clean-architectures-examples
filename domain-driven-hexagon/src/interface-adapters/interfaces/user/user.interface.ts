import { ModelBase } from '../../../libs/ddd/interface-adapters/interfaces/model.base.interface';

export interface User extends ModelBase {
  email: string;
  country: string;
  postalCode: string;
  street: string;
}
