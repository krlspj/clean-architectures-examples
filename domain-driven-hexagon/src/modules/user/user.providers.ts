import { Logger, Provider } from '@nestjs/common';

/* Constructing custom providers
 */
export const createUserCliLoggerSymbol = Symbol('createUserCliLoggerSymbol');

export const createUserCliLoggerProvider: Provider = {
  provide: createUserCliLoggerSymbol,
  useFactory: (): Logger => {
    return new Logger('create-user-cli');
  },
};
