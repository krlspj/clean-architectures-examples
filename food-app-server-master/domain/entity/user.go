package entity

import (
	"food-app/infrastructure/security"
	"github.com/badoux/checkmail"
	"html"
	"strings"
	"time"
)

type User struct {
	ID        uint64     `gorm:"primary_key;auto_increment" json:"id"`
	FirstName string     `gorm:"size:100;not null;" json:"first_name"`
	LastName  string     `gorm:"size:100;not null;" json:"last_name"`
	Email     string     `gorm:"size:100;not null;unique" json:"email"`
	Password  string     `gorm:"size:100;not null;" json:"password"`
	CreatedAt time.Time  `gorm:"default:CURRENT_TIMESTAMP" json:"created_at"`
	UpdatedAt time.Time  `gorm:"default:CURRENT_TIMESTAMP" json:"updated_at"`
	DeletedAt *time.Time `json:"deleted_at,omitempty"`
}

type PublicUser struct {
	ID        uint64 `gorm:"primary_key;auto_increment" json:"id"`
	FirstName string `gorm:"size:100;not null;" json:"first_name"`
	LastName  string `gorm:"size:100;not null;" json:"last_name"`
}

//BeforeSave is a gorm hook
func (u *User) BeforeSave() error {
	hashPassword, err := security.Hash(u.Password)
	if err != nil {
		return err
	}
	u.Password = string(hashPassword)
	return nil
}

type Users []User

//So that we dont expose the user's email address and password to the world
func (users Users) PublicUsers() []interface{} {
	result := make([]interface{}, len(users))
	for index, user := range users {
		result[index] = user.PublicUser()
	}
	return result
}

//So that we dont expose the user's email address and password to the world
func (u *User) PublicUser() interface{} {
	return &PublicUser{
		ID:        u.ID,
		FirstName: u.FirstName,
		LastName:  u.LastName,
	}
}

func (u *User) Prepare() {
	u.FirstName = html.EscapeString(strings.TrimSpace(u.FirstName))
	u.LastName = html.EscapeString(strings.TrimSpace(u.LastName))
	u.Email = html.EscapeString(strings.TrimSpace(u.Email))
	u.CreatedAt = time.Now()
	u.UpdatedAt = time.Now()
}

func (u *User) Validate(action string) map[string]string {
	var errorMessages = make(map[string]string)
	var err error

	switch strings.ToLower(action) {
	case "update":
		if u.Email == "" {
			errorMessages["email_required"] = "email required"
		}
		if u.Email != "" {
			if err = checkmail.ValidateFormat(u.Email); err != nil {
				errorMessages["invalid_email"] = "email email"
			}
		}

	case "login":
		if u.Password == "" {
			errorMessages["password_required"] = "password is required"
		}
		if u.Email == "" {
			errorMessages["email_required"] = "email is required"
		}
		if u.Email != "" {
			if err = checkmail.ValidateFormat(u.Email); err != nil {
				errorMessages["invalid_email"] = "please provide a valid email"
			}
		}
	case "forgotpassword":
		if u.Email == "" {
			errorMessages["email_required"] = "email required"
		}
		if u.Email != "" {
			if err = checkmail.ValidateFormat(u.Email); err != nil {
				errorMessages["invalid_email"] = "please provide a valid email"
			}
		}
	default:
		if u.FirstName == "" {
			errorMessages["firstname_required"] = "first name is required"
		}
		if u.LastName == "" {
			errorMessages["lastname_required"] = "last name is required"
		}
		if u.Password == "" {
			errorMessages["password_required"] = "password is required"
		}
		if u.Password != "" && len(u.Password) < 6 {
			errorMessages["invalid_password"] = "password should be at least 6 characters"
		}
		if u.Email == "" {
			errorMessages["email_required"] = "email is required"
		}
		if u.Email != "" {
			if err = checkmail.ValidateFormat(u.Email); err != nil {
				errorMessages["invalid_email"] = "please provide a valid email"
			}
		}
	}
	return errorMessages
}
