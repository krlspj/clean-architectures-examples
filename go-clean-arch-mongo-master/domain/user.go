package domain

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type User struct {
	ID        primitive.ObjectID `bson:"_id" json:"id"`
	CreatedAt time.Time          `bson:"created_at" json:"created_at"`
	UpdatedAt time.Time          `bson:"updated_at" json:"updated_at"`
	Name      string             `bson:"name" json:"name" validate:"required"`
	Username  string             `bson:"username" json:"username" validate:"required"`
	Password  string             `bson:"password" json:"password" validate:"required"`
}

type UserRepository interface {
	InsertOne(ctx context.Context, u *User) (*User, error)
	FindOne(ctx context.Context, id string) (*User, error)
	GetAllWithPage(ctx context.Context, rp int64, p int64, filter interface{}, setsort interface{}) ([]User, int64, error)
	UpdateOne(ctx context.Context, user *User, id string) (*User, error)
	GetByCredential(ctx context.Context, username string, password string) (*User, error)
}

type UserUsecase interface {
	InsertOne(ctx context.Context, u *User) (*User, error)
	FindOne(ctx context.Context, id string) (*User, error)
	GetAllWithPage(ctx context.Context, rp int64, p int64, filter interface{}, setsort interface{}) ([]User, int64, error)
	UpdateOne(ctx context.Context, user *User, id string) (*User, error)
}
