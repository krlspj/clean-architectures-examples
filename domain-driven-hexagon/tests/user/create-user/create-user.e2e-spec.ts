import { defineFeature, loadFeature } from 'jest-cucumber';
import * as request from 'supertest';
import { CreateUser } from '@src/interface-adapters/interfaces/user/create.user.interface';
import { Id } from '@src/libs/ddd/interface-adapters/interfaces/id.interface';
import { UserResponse } from '@src/modules/user/dtos/user.response.dto';
import { snapshotBaseProps } from '@src/libs/test-utils/snapshot-base-props';
import { getTestServer, TestServer } from '../../jestSetupAfterEnv';

const feature = loadFeature('tests/user/create-user/create-user.feature');

/**
 * e2e test implementing a Gherkin feature file
 * https://github.com/Sairyss/backend-best-practices#testing
 */

defineFeature(feature, test => {
  let testServer: TestServer;
  let httpServer: request.SuperTest<request.Test>;

  beforeAll(() => {
    testServer = getTestServer();
    httpServer = request(testServer.serverApplication.getHttpServer());
  });

  afterAll(() => {
    // TODO: clean db after tests are finished
  });

  test('Creating a user happy path', ({ given, when, then, and }) => {
    const userDto: Partial<CreateUser> = {};
    let userId: Id;

    given(/^that my email is "(.*)"$/, (email: string) => {
      userDto.email = email;
    });

    and(
      /^my country is "(.*)", my postal code is "(.*)" and my street is "(.*)"$/,
      (country: string, postalCode: string, street: string) => {
        userDto.country = country;
        userDto.postalCode = postalCode;
        userDto.street = street;
      },
    );

    when('I send a request to create a user', async () => {
      const res = await httpServer
        .post('/v1/users')
        .send(userDto)
        .expect(201);
      userId = res.body;
    });

    then('I receive my user ID', () => {
      expect(userId).toMatchSnapshot({ id: expect.any(String) });
    });

    and('I can see my user in a list of all users', async () => {
      const res = await httpServer.get('/v1/users').expect(200);

      expect(res.body).toMatchSnapshot([snapshotBaseProps]);
      expect(res.body.some((item: UserResponse) => item.id === userId.id)).toBe(
        true,
      );
    });
  });
});
