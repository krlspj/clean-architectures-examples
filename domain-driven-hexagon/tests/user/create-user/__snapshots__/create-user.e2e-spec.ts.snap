// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create a user Creating a user happy path 1`] = `
Object {
  "id": Any<String>,
}
`;

exports[`Create a user Creating a user happy path 2`] = `
Array [
  Object {
    "country": "England",
    "createdAt": Any<String>,
    "email": "john.doe@gmail.com",
    "id": Any<String>,
    "postalCode": "29145",
    "street": "Road Avenue",
    "updatedAt": Any<String>,
  },
]
`;
